protocol/claim_monero.go

```diff
@@ -0,0 +1,115 @@
+package protocol
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/athanorlabs/atomic-swap/coins"
+	"github.com/athanorlabs/atomic-swap/common"
+	"github.com/athanorlabs/atomic-swap/common/types"
+	mcrypto "github.com/athanorlabs/atomic-swap/crypto/monero"
+	"github.com/athanorlabs/atomic-swap/monero"
+
+	logging "github.com/ipfs/go-log"
+)
+
+var (
+	log = logging.Logger("protocol")
+)
+
+// GetClaimKeypair returns the private key pair required for a monero claim.
+// The key pair is the summation of each party's private spend and view keys:
+// (s_a + s_b) and (v_a + v_b).
+func GetClaimKeypair(
+	skA, skB *mcrypto.PrivateSpendKey,
+	vkA, vkB *mcrypto.PrivateViewKey,
+) *mcrypto.PrivateKeyPair {
+	skAB := mcrypto.SumPrivateSpendKeys(skA, skB)
+	vkAB := mcrypto.SumPrivateViewKeys(vkA, vkB)
+	kpAB := mcrypto.NewPrivateKeyPair(skAB, vkAB)
+	return kpAB
+}
+
+// ClaimMonero claims the XMR located in the wallet controlled by the private keypair `kpAB`.
+// If transferBack is true, it sweeps the XMR to `depositAddr`.
+func ClaimMonero(
+	ctx context.Context,
+	env common.Environment,
+	id types.Hash,
+	xmrClient monero.WalletClient,
+	walletScanHeight uint64,
+	kpAB *mcrypto.PrivateKeyPair,
+	depositAddr mcrypto.Address,
+	transferBack bool,
+) (mcrypto.Address, error) {
+	abAddr := kpAB.PublicKeyPair().Address(env)
+
+	conf := xmrClient.CreateWalletConf(fmt.Sprintf("swap-wallet-claim-%s", id))
+	abWalletCli, err := monero.CreateSpendWalletFromKeys(conf, kpAB, walletScanHeight)
+	if err != nil {
+		return "", err
+	}
+
+	if transferBack {
+		defer abWalletCli.CloseAndRemoveWallet()
+	} else {
+		abWalletCli.Close()
+		log.Infof("monero claimed in account %s with wallet file %s", abAddr, conf.WalletFilePath)
+		return abAddr, nil
+	}
+
+	log.Infof("monero claimed in account %s; transferring to deposit account %s",
+		abAddr, depositAddr)
+
+	err = mcrypto.ValidateAddress(string(depositAddr), env)
+	if err != nil {
+		log.Errorf(
+			"failed to transfer XMR out of swap wallet, dest address %s is invalid: %s",
+			abAddr,
+			err,
+		)
+		return "", err
+	}
+
+	err = waitUntilBalanceUnlocks(ctx, abWalletCli)
+	if err != nil {
+		return "", fmt.Errorf("failed to wait for balance to unlock: %w", err)
+	}
+
+	transfers, err := abWalletCli.SweepAll(ctx, depositAddr, 0, monero.SweepToSelfConfirmations)
+	if err != nil {
+		return "", fmt.Errorf("failed to send funds to deposit account: %w", err)
+	}
+
+	for _, transfer := range transfers {
+		log.Infof("transferred %s XMR to primary wallet (%s XMR lost to fees)",
+			coins.FmtPiconeroAmtAsXMR(transfer.Amount),
+			coins.FmtPiconeroAmtAsXMR(transfer.Fee),
+		)
+	}
+
+	return abAddr, nil
+}
+
+// TODO: Put this in monero package? Unit test.
```

Comment:
If it is only used by the `ClaimMonero` function above, here is probably the best place (easiest to find for someone reviewing the code). Its reuse potential is only with AB swap wallets, as the primary wallet's funds may be in a constant state of flux, were some portion of its funds are always locked.